In [46]:
#perform naive bayes, logistic regression, and decision tree classification for online fraud dataset
#target or predicted column : is fraud...... which one of three methods has highest accuracy


In [101]:
import pandas as pd
from sklearn import metrics
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import export_graphviz
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

In [48]:
df = pd.read_csv('onlinefraud.csv')

In [49]:
df

step      type      amount     nameOrig  oldbalanceOrg  \
0           1   PAYMENT     9839.64  C1231006815      170136.00   
1           1   PAYMENT     1864.28  C1666544295       21249.00   
2           1  TRANSFER      181.00  C1305486145         181.00   
3           1  CASH_OUT      181.00   C840083671         181.00   
4           1   PAYMENT    11668.14  C2048537720       41554.00   
...       ...       ...         ...          ...            ...   
6362615   743  CASH_OUT   339682.13   C786484425      339682.13   
6362616   743  TRANSFER  6311409.28  C1529008245     6311409.28   
6362617   743  CASH_OUT  6311409.28  C1162922333     6311409.28   
6362618   743  TRANSFER   850002.52  C1685995037      850002.52   
6362619   743  CASH_OUT   850002.52  C1280323807      850002.52   

         newbalanceOrig     nameDest  oldbalanceDest  newbalanceDest  isFraud  \
0             160296.36  M1979787155            0.00            0.00        0   
1              19384.72  M2044282225            0.00            0.00        0   
2                  0.00   C553264065            0.00            0.00        1   
3                  0.00    C38997010        21182.00            0.00        1   
4              29885.86  M1230701703            0.00            0.00        0   
...                 ...          ...             ...             ...      ...   
6362615            0.00   C776919290            0.00       339682.13        1   
6362616            0.00  C1881841831            0.00            0.00        1   
6362617            0.00  C1365125890        68488.84      6379898.11        1   
6362618            0.00  C2080388513            0.00            0.00        1   
6362619            0.00   C873221189      6510099.11      7360101.63        1   

         isFlaggedFraud  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  
...                 ...  
6362615               0  
6362616               0  
6362617               0  
6362618               0  
6362619               0  

[6362620 rows x 11 columns]

In [50]:
df.dtypes

step                int64
type               object
amount            float64
nameOrig           object
oldbalanceOrg     float64
newbalanceOrig    float64
nameDest           object
oldbalanceDest    float64
newbalanceDest    float64
isFraud             int64
isFlaggedFraud      int64
dtype: object

In [51]:
features = ['type','amount','nameOrig','oldbalanceOrg','newbalanceOrig','oldbalanceDest','newbalanceDest']
target = 'isFraud'

In [52]:
#Encoding
#Encoding 
encoder = LabelEncoder()
for col in features:
    if df[col].dtype == object:
        df[col] = encoder.fit_transform(df[col])
df

step  type      amount  nameOrig  oldbalanceOrg  newbalanceOrig  \
0           1     3     9839.64    757869      170136.00       160296.36   
1           1     3     1864.28   2188998       21249.00        19384.72   
2           1     4      181.00   1002156         181.00            0.00   
3           1     1      181.00   5828262         181.00            0.00   
4           1     3    11668.14   3445981       41554.00        29885.86   
...       ...   ...         ...       ...            ...             ...   
6362615   743     1   339682.13   5651847      339682.13            0.00   
6362616   743     4  6311409.28   1737278     6311409.28            0.00   
6362617   743     1  6311409.28    533958     6311409.28            0.00   
6362618   743     4   850002.52   2252932      850002.52            0.00   
6362619   743     1   850002.52    919229      850002.52            0.00   

            nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0        M1979787155            0.00            0.00        0               0  
1        M2044282225            0.00            0.00        0               0  
2         C553264065            0.00            0.00        1               0  
3          C38997010        21182.00            0.00        1               0  
4        M1230701703            0.00            0.00        0               0  
...              ...             ...             ...      ...             ...  
6362615   C776919290            0.00       339682.13        1               0  
6362616  C1881841831            0.00            0.00        1               0  
6362617  C1365125890        68488.84      6379898.11        1               0  
6362618  C2080388513            0.00            0.00        1               0  
6362619   C873221189      6510099.11      7360101.63        1               0  

[6362620 rows x 11 columns]

In [53]:
#Scaling
scaler = StandardScaler()
df[['amount','nameOrig','oldbalanceOrg','newbalanceOrig','oldbalanceDest','newbalanceDest']]= scaler.fit_transform(df[['amount','nameOrig','oldbalanceOrg','newbalanceOrig','oldbalanceDest','newbalanceDest']])

In [54]:
df

step  type     amount  nameOrig  oldbalanceOrg  newbalanceOrig  \
0           1     3  -0.281560 -1.318825      -0.229810       -0.237622   
1           1     3  -0.294767 -0.538520      -0.281359       -0.285812   
2           1     4  -0.297555 -1.185630      -0.288654       -0.292442   
3           1     1  -0.297555  1.445742      -0.288654       -0.292442   
4           1     3  -0.278532  0.146834      -0.274329       -0.282221   
...       ...   ...        ...       ...            ...             ...   
6362615   743     1   0.264665  1.349554      -0.171108       -0.292442   
6362616   743     4  10.153953 -0.784815       1.896491       -0.292442   
6362617   743     1  10.153953 -1.440909       1.896491       -0.292442   
6362618   743     4   1.109765 -0.503661       0.005581       -0.292442   
6362619   743     1   1.109765 -1.230845       0.005581       -0.292442   

            nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0        M1979787155       -0.323814       -0.333411        0               0  
1        M2044282225       -0.323814       -0.333411        0               0  
2         C553264065       -0.323814       -0.333411        1               0  
3          C38997010       -0.317582       -0.333411        1               0  
4        M1230701703       -0.323814       -0.333411        0               0  
...              ...             ...             ...      ...             ...  
6362615   C776919290       -0.323814       -0.240959        1               0  
6362616  C1881841831       -0.323814       -0.333411        1               0  
6362617  C1365125890       -0.303665        1.403027        1               0  
6362618  C2080388513       -0.323814       -0.333411        1               0  
6362619   C873221189        1.591383        1.669812        1               0  

[6362620 rows x 11 columns]

In [55]:
X = df[features]
y = df[target]

In [56]:
X_train,X_test,Y_train,Y_test= train_test_split(X,y,test_size=0.2,random_state=1)

# Prediction through Naive bayes theorm

In [58]:

nb = GaussianNB()
nb.fit(X_train,Y_train)

GaussianNB()

In [72]:
y_pred = nb.predict(X_test)

In [74]:
y_pred

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [82]:
#Print various classification metrics
print('Confusion Matrix : ', metrics.confusion_matrix(Y_test, y_pred))
print('Accuracy: ', metrics.accuracy_score(Y_test, y_pred))
print('Precision Score: ', metrics.precision_score(Y_test, y_pred))
print('Recall  Score: ', metrics.recall_score(Y_test, y_pred))
print('F1 Score: ', metrics.f1_score(Y_test, y_pred))
print('Classification Report: ', metrics.classification_report(Y_test, y_pred))


Confusion Matrix :  [[1262633    8244]
 [   1385     262]]
Accuracy:  0.9924331486085921
Precision Score:  0.030801786973900777
Recall  Score:  0.15907710989678203
F1 Score:  0.0516103614695164
Classification Report:                precision    recall  f1-score   support

           0       1.00      0.99      1.00   1270877
           1       0.03      0.16      0.05      1647

    accuracy                           0.99   1272524
   macro avg       0.51      0.58      0.52   1272524
weighted avg       1.00      0.99      0.99   1272524



# Prediction through Logistic Regression

In [109]:
model_lr = LogisticRegression()
model_lr.fit(X_train,Y_train) 

LogisticRegression()

In [115]:
y_pred_lr = model_lr.predict(X_test)

In [117]:
#Print various classification metrics
print('Confusion Matrix : ', metrics.confusion_matrix(Y_test, y_pred_lr))
print('Accuracy: ', metrics.accuracy_score(Y_test, y_pred_lr))
print('Precision Score: ', metrics.precision_score(Y_test, y_pred_lr))
print('Recall  Score: ', metrics.recall_score(Y_test, y_pred_lr))
print('F1 Score: ', metrics.f1_score(Y_test, y_pred_lr))
print('Classification Report: ', metrics.classification_report(Y_test, y_pred_lr))

Confusion Matrix :  [[1270791      86]
 [   1006     641]]
Accuracy:  0.9991418629432529
Precision Score:  0.8817056396148556
Recall  Score:  0.38919247115968425
F1 Score:  0.540016849199663
Classification Report:                precision    recall  f1-score   support

           0       1.00      1.00      1.00   1270877
           1       0.88      0.39      0.54      1647

    accuracy                           1.00   1272524
   macro avg       0.94      0.69      0.77   1272524
weighted avg       1.00      1.00      1.00   1272524



# Prediction through Decision Tree

In [123]:
modeltr = DecisionTreeClassifier()
modeltr = modeltr.fit(X_train , Y_train)
modeltr

DecisionTreeClassifier()

In [126]:
y_predtr  = model.predict(X_test)

In [127]:
#Print various classification metrics
print('Confusion Matrix : ', metrics.confusion_matrix(Y_test, y_pred_lr))
print('Accuracy: ', metrics.accuracy_score(Y_test, y_pred_lr))
print('Precision Score: ', metrics.precision_score(Y_test, y_pred_lr))
print('Recall  Score: ', metrics.recall_score(Y_test, y_pred_lr))
print('F1 Score: ', metrics.f1_score(Y_test, y_pred_lr))
print('Classification Report: ', metrics.classification_report(Y_test, y_pred_lr))

Confusion Matrix :  [[1270791      86]
 [   1006     641]]
Accuracy:  0.9991418629432529
Precision Score:  0.8817056396148556
Recall  Score:  0.38919247115968425
F1 Score:  0.540016849199663
Classification Report:                precision    recall  f1-score   support

           0       1.00      1.00      1.00   1270877
           1       0.88      0.39      0.54      1647

    accuracy                           1.00   1272524
   macro avg       0.94      0.69      0.77   1272524
weighted avg       1.00      1.00      1.00   1272524



In [128]:
#Print importance of each feature
print('Feature Importance')
for feature, importance in zip(features, model.feature_importances_):
    print(f'{feature} :{importance :.2f} ')

Feature Importance
type :0.02 
amount :0.15 
nameOrig :0.02 
oldbalanceOrg :0.41 
newbalanceOrig :0.08 
oldbalanceDest :0.08 
newbalanceDest :0.23 


In [135]:
#Save the tree as image 
#Three options
#OPtion 1 - Create dot file and then convert manually to png format
#Save the tree as DOT file 
with open('OnlineFraud.dot' , 'w') as f:
    export_graphviz(model , out_file = f, feature_names=features, filled = True)
#You have to say dot -Tpng titanic_tree.dot -o titanic_tree.png to convert 


# Conclusion

# Accuracy of Naive Bayes:  0.9924331486085921
# Accuracy of Logistic Regression:  0.9991418629432529
# Accuracy of decision Tree:  0.9991418629432529